optuna/samplers/_nsga2/crossover.py

```diff
@@ -0,0 +1,456 @@
+        else:
+            assert False
+
+        _params = [
+            trans.untransform(np.array([param])) for trans, param in zip(transes, params_array)
+        ]
+        for param in _params:
+            for param_name in param.keys():
+                child[param_name] = param[param_name]
+
+        if _is_contained(child, search_space):
+            break
+
+    return child
+
+
+def _selection(
+    crossover_name: str,
+    study: Study,
+    parent_population: Sequence[FrozenTrial],
+    rng: np.random.RandomState,
+    dominates: Callable[[FrozenTrial, FrozenTrial, Sequence[StudyDirection]], bool],
+) -> List[FrozenTrial]:
+    if crossover_name in ["uniform", "blxalpha", "sbx", "vsbx"]:
+        n_select = 2
+    elif crossover_name in ["undx", "spx"]:
+        n_select = 3
+    elif crossover_name in ["undxm"]:
+        n_select = 4
+    else:
+        assert False
+    if len(parent_population) < n_select:
+        raise ValueError(
+            f"Using {crossover_name},"
+            f"the population size should be greater than or equal to {n_select}."
+        )
+
+    parents = []
+    for _ in range(n_select):
+        parent = _select_parent(
+            study, [t for t in parent_population if t not in parents], rng, dominates
+        )
+        parents.append(parent)
+    return parents
+
+
+def _select_parent(
+    study: Study,
+    population: Sequence[FrozenTrial],
+    rng: np.random.RandomState,
+    dominates: Callable[[FrozenTrial, FrozenTrial, Sequence[StudyDirection]], bool],
+) -> FrozenTrial:
+    # TODO(ohta): Consider to allow users to specify the number of parent candidates.
+    population_size = len(population)
+    candidate0 = population[rng.choice(population_size)]
+    candidate1 = population[rng.choice(population_size)]
+
+    # TODO(ohta): Consider crowding distance.
+    if dominates(candidate0, candidate1, study.directions):
+        return candidate0
+    else:
+        return candidate1
+
+
+def _swap(p0_i: Any, p1_i: Any, rand: float, swapping_prob: float) -> Any:
+    if rand < swapping_prob:
+        return p1_i
+    else:
+        return p0_i
+
+
+def _uniform(
+    x1: np.ndarray, x2: np.ndarray, rng: np.random.RandomState, swapping_prob: float
+) -> np.ndarray:
+    # https://www.researchgate.net/publication/201976488_Uniform_Crossover_in_Genetic_Algorithms
+    # Section 1 Introduction
+
+    child = []
+    for x1_i, x2_i in zip(x1, x2):
+        param = _swap(x1_i, x2_i, rng.rand(), swapping_prob)
+        child.append(param)
+    return np.array(child)
+
+
+def _blxalpha(
+    x1: np.ndarray, x2: np.ndarray, rng: np.random.RandomState, alpha: float
+) -> np.ndarray:
+    # https://www.sciencedirect.com/science/article/abs/pii/B9780080948324500180
+
+    # http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.465.6900&rep=rep1&type=pdf
+    # Section 2 Crossover Operators for RCGA 2.1 Blend Crossover
+
+    assert x1.shape == x2.shape
```

Comment:
Do we have any other assumptions on the shape of input? For example, `len(x1.shape) == 1`?

Reply:
This comment will be applied in other functions.